helper/cidrutil/cidr.go

```diff
@@ -0,0 +1,200 @@
+// the given CIDR blocks, when the input is a string composed by joining all
+// the CIDR blocks using a separator. The input is separated based on the given
+// separator and the IP is checked to be belonged by any CIDR block.
+func IPBelongsToCIDRBlocksString(ipAddr string, cidrList, separator string) (bool, error) {
+	if ipAddr == "" {
+		return false, fmt.Errorf("missing IP address")
+	}
+
+	if cidrList == "" {
+		return false, fmt.Errorf("missing CIDR list")
+	}
+
+	if separator == "" {
+		return false, fmt.Errorf("missing separator")
+	}
+
+	if ip := net.ParseIP(ipAddr); ip == nil {
+		return false, fmt.Errorf("invalid IP address")
+	}
+
+	return IPBelongsToCIDRBlocksSlice(ipAddr, strutil.ParseDedupAndSortStrings(cidrList, separator))
+}
+
+// IPBelongsToCIDRBlocksSlice checks if the given IP is encompassed by any of the given
+// CIDR blocks
+func IPBelongsToCIDRBlocksSlice(ipAddr string, cidrs []string) (bool, error) {
+	if ipAddr == "" {
+		return false, fmt.Errorf("missing IP address")
+	}
+
+	if len(cidrs) == 0 {
+		return false, fmt.Errorf("missing CIDR blocks to be checked against")
+	}
+
+	if ip := net.ParseIP(ipAddr); ip == nil {
+		return false, fmt.Errorf("invalid IP address")
+	}
+
+	for _, cidr := range cidrs {
+		belongs, err := IPBelongsToCIDR(ipAddr, cidr)
+		if err != nil {
+			return false, err
+		}
+		if belongs {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// ValidateCIDRListString checks if the list of CIDR blocks are valid, given
+// that the input is a string composed by joining all the CIDR blocks using a
+// separator. The input is separated based on the given separator and validity
+// of each is checked.
+func ValidateCIDRListString(cidrList string, separator string) (bool, error) {
+	if cidrList == "" {
+		return false, fmt.Errorf("missing CIDR list that needs validation")
+	}
+	if separator == "" {
+		return false, fmt.Errorf("missing separator")
+	}
+
+	return ValidateCIDRListSlice(strutil.ParseDedupAndSortStrings(cidrList, separator))
+}
+
+// ValidateCIDRListSlice checks if the given list of CIDR blocks are valid
+func ValidateCIDRListSlice(cidrBlocks []string) (bool, error) {
+	if len(cidrBlocks) == 0 {
+		return false, fmt.Errorf("missing CIDR blocks that needs validation")
+	}
+
+	for _, block := range cidrBlocks {
+		if _, _, err := net.ParseCIDR(strings.TrimSpace(block)); err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+// Subset checks if the IPs belonging to a given CIDR block is a subset of IPs
+// belonging to another CIDR block.
+func Subset(cidr1, cidr2 string) (bool, error) {
+	if cidr1 == "" {
+		return false, fmt.Errorf("missing CIDR to be checked against")
+	}
+
+	if cidr2 == "" {
+		return false, fmt.Errorf("missing CIDR that needs to be checked")
+	}
+
+	_, net1, err := net.ParseCIDR(cidr1)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse the CIDR to be checked against: %q", err)
+	}
+
+	maskLen1, _ := net1.Mask.Size()
+	if maskLen1 == 0 {
```

Comment:
The case of mask length being zero, is not just for zero address. Looking at the source code, it returns zero when the CIDR is not in its canonical form. I'll have to special case the zero address. Thanks for pointing this out.